xwiki-platform-core/xwiki-platform-notifications/xwiki-platform-notifications-filters/xwiki-platform-notifications-filters-api/src/main/java/org/xwiki/notifications/filters/internal/scope/ScopeNotificationFilterLocationStateComputer.java

```diff
@@ -153,31 +218,55 @@ private WatchedLocationState handleExclusiveFilters(EntityReference location,
             ScopeNotificationFilterPreference pref = it.next();
 
             int deepLevel = pref.getScopeReference().size();
+            boolean isExactMatch = isExactMatch(pref, location);
+            boolean isParentMatch = isParentMatch(pref, location);
+            boolean isSpaceMatch = isSpaceMatch(pref, location);
 
             // If the exclusive filter match the event location...
-            if (match(pref, location) && deepLevel > deepestLevel) {
-                state = WatchedState.NOT_WATCHED;
+            if ((isExactMatch || isParentMatch) && deepLevel > deepestLevel) {
+                state = getWatchedState(allTypesAndEvents, isExactMatch, isSpaceMatch, false);
                 startingDate = pref.getStartingDate();
                 deepestLevel = deepLevel;
 
                 // then we watch the location if there is at least an inclusive filter child matching it
                 for (ScopeNotificationFilterPreference child : pref.getChildren()) {
                     int childDeepLevel = child.getScopeReference().size();
-                    if (match(child, location) && childDeepLevel > deepestLevel) {
-                        state = WatchedState.WATCHED;
+                    boolean isChildExactMatch = isExactMatch(child, location);
+                    boolean isChildParentMatch = isParentMatch(child, location);
+                    boolean isChildSpaceMatch = isSpaceMatch(child, location);
+                    if ((isChildExactMatch || isChildParentMatch) && childDeepLevel > deepestLevel) {
+                        state = getWatchedState(allTypesAndEvents, isChildExactMatch, isChildSpaceMatch, true);
                         deepestLevel = childDeepLevel;
                         startingDate = child.getStartingDate();
                     }
                 }
             }
         }
 
-        return state == WatchedState.UNKNOWN ? null
-            : new WatchedLocationState(state == WatchedState.WATCHED, startingDate);
+        if (state != null) {
+            return Optional.of(new WatchedLocationState(state, startingDate));
+        } else {
+            return Optional.empty();
+        }
```

Comment:
FYI, `Optional.ofNullable(state).map(s -> new WatchedLocationState(s, startingDate))` can spare you an if/else block and is more idiomatic.
But it's also a matter of taste :)

Reply:
I personally find `Optional.ofNullable(state).map(s -> new WatchedLocationState(s, startingDate))` a lot more complex to read but it's probably because I'm not used to it (whereas reading an IF is very simple for me).

Reply:
> I personally find Optional.ofNullable(state).map(s -> new WatchedLocationState(s, startingDate)) a lot more complex to read but it's probably because I'm not used to it (whereas reading an IF is very simple for me).

It's same for me.